sematic/abstract_future.py

```diff
@@ -239,6 +279,19 @@ def props(self) -> FutureProperties:
         """
         return self._props
 
+    @property
+    def state(self) -> FutureState:
+        return self._props.state
+
+    @state.setter
+    def state(self, state: FutureState) -> None:
+        self._props.state = state
+        if state is FutureState.SCHEDULED and self.props.scheduled_epoch_time is None:
+            self._props.scheduled_epoch_time = time.time()
+        if state.is_terminal() or state is FutureState.RETRYING:
+            # ensures that timeout resets for retries.
+            self._props.scheduled_epoch_time = None
```

Comment:
Is there a chance that this captures overhead time into the timeout as well? This notably captures the time to spin up the pod, download the image, etc, all of which can easily take over a minute.

Reply:
Yes, it includes the time to spin up the pod and start work. (1) The rest of our code doesn't distinguish those states (we just have one "scheduled" state for both the pending and the running) (2) I think it's actually somewhat desirable, as if something sits pending for too long it will timeout and fail rather than just being indefinitely hung.

Reply:
Additionally, given the fact that timeouts on outer functions apply to the time it takes inner functions to execute, there's really no way to avoid some "pending" time contributing towards the timeout. This at least makes it consistent.